src/execution.ts

```diff
@@ -1637,3 +1669,194 @@ function getParentArgIndexes(context: CompilationContext) {
+ * executeSubscription.
+ */
+
+function isAsyncIterable<T = unknown>(
+  val: unknown
+): val is AsyncIterableIterator<T> {
+  return typeof Object(val)[Symbol.asyncIterator] === "function";
+}
+
+function compileSubscriptionOperation(
+  context: CompilationContext,
+  type: GraphQLObjectType,
+  fieldMap: FieldsAndNodes,
+  queryFn: CompiledQuery["query"]
+) {
+  function reportGraphQLError(error: any): ExecutionResult {
+    if (error instanceof GraphQLError) {
+      return {
+        errors: [error]
+      };
+    }
+    throw error;
+  }
+
+  const fieldNodes = Object.values(fieldMap)[0];
+  const fieldNode = fieldNodes[0];
+  const fieldName = fieldNode.name.value;
+
+  const field = resolveFieldDef(context, type, fieldNodes);
+
+  if (!field) {
+    throw new GraphQLError(
+      `The subscription field "${fieldName}" is not defined.`,
+      fieldNodes
+    );
+  }
+
+  const responsePath = addPath(undefined, fieldName);
+  const resolveInfoName = createResolveInfoName(responsePath);
+
+  // Call the `subscribe()` resolver or the default resolver to produce an
+  // AsyncIterable yielding raw payloads.
+  const subscriber = field.subscribe;
+
+  return async function subscribe(executionContext: ExecutionContext) {
+    const resolveInfo = executionContext.resolveInfos[resolveInfoName](
+      executionContext.rootValue,
+      executionContext.variables,
+      responsePath
+    );
+
+    let resultOrStream: AsyncIterableIterator<any>;
+
+    try {
+      try {
+        resultOrStream =
+          subscriber &&
+          (await subscriber(
+            executionContext.rootValue,
+            executionContext.variables,
+            executionContext.context,
+            resolveInfo
+          ));
+        if (resultOrStream instanceof Error) {
+          throw resultOrStream;
+        }
+      } catch (error) {
+        throw locatedError(
+          error,
+          resolveInfo.fieldNodes,
+          pathToArray(resolveInfo.path)
+        );
+      }
+      if (!isAsyncIterable(resultOrStream)) {
+        throw new Error(
+          "Subscription field must return Async Iterable. " +
+            `Received: ${inspect(resultOrStream)}.`
+        );
+      }
+    } catch (e) {
+      return reportGraphQLError(e);
+    }
+
+    // For each payload yielded from a subscription, map it over the normal
+    // GraphQL `execute` function, with `payload` as the rootValue.
+    // This implements the "MapSourceToResponseEvent" algorithm described in
+    // the GraphQL specification. The `execute` function provides the
+    // "ExecuteSubscriptionEvent" algorithm, as it is nearly identical to the
+    // "ExecuteQuery" algorithm, for which `execute` is also used.
+    // We use our `query` function in place of `execute`
+    const mapSourceToResponse = (payload: any) =>
+      queryFn(payload, executionContext.context, executionContext.variables);
+
+    return mapAsyncIterator(
```

Comment:
Probably an unimportant issue - the arguments to the `next()` fn in the async iterator is ignored in the implementation of the function `mapAsyncIterator` - 

https://github.com/graphql/graphql-js/blob/fe9ea6db90a12c88d6372f88b61ca68e7aa3659b/src/subscription/mapAsyncIterator.js#L42-L43

I was trying to pass back a value from the GraphQL subscription result to the async generator function - 

```ts
// implementation
async *subscribe() {
  let previousValue = initialValue ?? previousValueFromInput;
  for (;;) {
    previousValue = yield promise(previousValue);
  }
}
```

```ts
// usage
const sub = await subscribe(`subscription { value }`);
const firstValue = await sub.next(
  something, client, wants, to, send, during, iteration /* These arguments are ignored */
);
```

This is probably a GraphQL-JS bug, I'm just curious if there are no usages of this pattern. We do not have to resolve this comment in this PR.

Reply:
That is indeed a valid pattern. Perhaps we can implement our own version of mapAsyncIterator?

Reply:
Yup mapAsyncIterator was changed to named export:

https://github.com/APIs-guru/graphql-js/blob/28315cd3f51b5ac212284f8583fb481053fa31a0/src/subscription/mapAsyncIterator.js